### PR TITLE
New Guide for Spring Cloud Stream

### PR DESCRIPTION
**Pull request by robertmcnees**
I noticed that our guide for Spring Cloud Stream lacked any code samples and only pointed to other documentation.  I added 3 new applications and documentation to walk a user through using various functionality for Spring Cloud Stream.

### CODE REVIEW

**Comment by robertmcnees** [comment]
Addresses #6

File: README.adoc [hunk 1]
```diff
@@ -2,25 +2,151 @@
...
 the framework from the link:https://spring.io/projects/spring-cloud-stream[project-site],
 link:https://spring.io/projects/spring-cloud-stream#learn[documentation],
 and link:https://github.com/spring-cloud/spring-cloud-stream-samples[samples].
 
-=== Stream Processing with Apache Kafka
-In the following guide, we develop three Spring Boot applications that use Spring Cloud Stream's support for Apache Kafka and deploy
-them to Cloud Foundry, to Kubernetes, and on your local machine.
+=== Stream Processing Application Overview
+
+In this guide, we will demonstrate the capability of Spring Cloud Stream.  We will create 3 applications to showcase the different
```
Comment (Buzzardo) [inline README.adoc:30]: Change to:

This guide demonstrates the capability of Spring Cloud Stream.  We create three applications to showcase the different

File: README.adoc [hunk 2]
```diff
@@ -2,25 +2,151 @@
...
+functionalities of Spring Cloud Stream.
+
+- name-source : publish a String at a regular interval to initiate the stream.  For this example, publish a name as a String.
```
Comment (Buzzardo) [inline README.adoc:33]: Remove the space before the colon. Capitalize the word after the colon. Use only one space before a period.

This comment applies to all the items in this list.

File: README.adoc [hunk 3]
```diff
@@ -2,25 +2,151 @@
...
+- name-processor : consume the String published by name-source and transform the data in some way.  Publish the results to a different exchange.
+For this example, use the name String to create a record and add a timestamp.
+- name-sink : consume the results from name-processor and perform an action.  In this case, print out the results to standard out.
```
Comment (Buzzardo) [inline README.adoc:36]: Remove "out" from "print out". It's unnecessary.

File: README.adoc [hunk 4]
```diff
@@ -2,25 +2,151 @@
...
+
+For this example the names of the applications follow the Spring Cloud Stream concepts (`Source`, `Processor`, `Sink`).  These concepts map
```
Comment (Buzzardo) [inline README.adoc:38]: Add a comma after "example". Use only one space after a period.

File: README.adoc [hunk 5]
```diff
@@ -2,25 +2,151 @@
...
+to the logical equivalent of the Java 8 functions (`Supplier`, Function, Consumer respectively).
```
Comment (Buzzardo) [inline README.adoc:39]: Add a comma before "respectively".

File: README.adoc [hunk 6]
```diff
@@ -2,25 +2,151 @@
...
+While Spring Cloud Stream can support a Function(s) in a Source  as well as Sink via function composition, we will have 3 separate applications to demonstrate how each can work as an independent application.
```
Comment (Buzzardo) [inline README.adoc:40]: Change to: 

While Spring Cloud Stream can support one or more Function instances in a Source  as well as in a Sink (via function composition), we have three separate applications to demonstrate how each can work as an independent application.

File: README.adoc [hunk 7]
```diff
@@ -2,25 +2,151 @@
...
+
+image::stream-applications-layered-architecture.png[]
+
+In this guide, we will work from the back to the front.  That is, we will first build the Sink application, followed by the Processor, and
```
Comment (Buzzardo) [inline README.adoc:44]: Remove all instances of "will". Use only one space after a period.

File: README.adoc [hunk 8]
```diff
@@ -2,25 +2,151 @@
...
+finally build the Source.  We will use the RabbitMQ dashboard UI to test each component as we go.
```
Comment (Buzzardo) [inline README.adoc:45]: Use only one space after a period.

File: README.adoc [hunk 9]
```diff
@@ -2,25 +2,151 @@
...
+
+=== Install Prerequisite - RabbitMQ
+
+In order to use Spring Cloud Stream functionality, we need to ensure that a message broker is accessible.  For this guide we will use RabbitMQ.
```
Comment (Buzzardo) [inline README.adoc:49]: Change to:

To use Spring Cloud Stream functionality, we need to ensure that a message broker is accessible.  For this guide, we use RabbitMQ.

File: README.adoc [hunk 10]
```diff
@@ -2,25 +2,151 @@
...
+If a local Docker environment is found, the following command can start RabbitMQ
```
Comment (Buzzardo) [inline README.adoc:50]: Add a colon at the end of this line.

File: README.adoc [hunk 11]
```diff
@@ -2,25 +2,151 @@
...
+```shell
+docker run -d --hostname my-rabbit --name some-rabbit -p 15672:15672 -p 5672:5672 rabbitmq:3-management
+```
+
+The result is that RabbitMQ should be link:http://localhost:15672/[accessible locally] with the username/password of guest/guest.
```
Comment (Buzzardo) [inline README.adoc:55]: Mark "guest/guest" as code (because the reader needs to type exactly those values).

File: README.adoc [hunk 12]
```diff
@@ -2,25 +2,151 @@
...
+
+
+== Sink Application
+
+The sink (`java.util.function.Consumer`) is defined in `NameSinkConfiguration` as
```
Comment (Buzzardo) [inline README.adoc:60]: Add a colon at the end of this line.

File: README.adoc [hunk 13]
```diff
@@ -2,25 +2,151 @@
...
+
+====
+[source,java,tabsize=2,indent=0]
+----
+include::name-sink/src/main/java/com/example/stream/namesink/NameSinkConfiguration.java[tag=namesink]
+----
+====
+
+Starting this application without configuring the exchange name will automatically generate an exchange in RabbitMQ titled `nameSink-in-0`.
+We want to customize this so that we can later hook our processor to our sink.
```
Comment (Buzzardo) [inline README.adoc:70]: Add "exchange" after "this".

File: README.adoc [hunk 14]
```diff
@@ -2,25 +2,151 @@
...
+The sink (`java.util.function.Consumer`) is defined in `NameSinkConfiguration` as
+
+====
+[source,java,tabsize=2,indent=0]
+----
+include::name-sink/src/main/java/com/example/stream/namesink/NameSinkConfiguration.java[tag=namesink]
+----
+====
+
+Starting this application without configuring the exchange name will automatically generate an exchange in RabbitMQ titled `nameSink-in-0`.
```
Comment (Buzzardo) [inline README.adoc:69]: Change to:

Starting this application without configuring the exchange name automatically generates an exchange named `nameSink-in-0` in RabbitMQ.

File: README.adoc [hunk 15]
```diff
@@ -2,25 +2,151 @@
...
+
+====
+[source,tabsize=2,indent=0]
+----
+include::name-sink/src/main/resources/application.properties[tag=sinkexchangeconfig]
+----
+====
+
+
+To test the sink we will manually publish a JSON message representing a `Person` record to the newly generated exchange:
```
Comment (Buzzardo) [inline README.adoc:80]: Add a comma after "To test the sink". Remove "will".

File: README.adoc [hunk 16]
```diff
@@ -2,25 +2,151 @@
...
+
+image::rabbitmq-sink.png[]
+
+And in our application logs we should see the data outputted:
```
Comment (Buzzardo) [inline README.adoc:84]: Change "And" to "Also," (including the comma).

(It's not strictly ungrammatical to start a sentence with a preposition - people do it all the time. However, technical documentation is formal, so we avoid that.)

File: README.adoc [hunk 17]
```diff
@@ -2,25 +2,151 @@
...
+
+image::output-sink.png[]
+
+
+
+== Processor Application
+
+The processor (`java.util.function.Function`) is defined in `NameProcessorConfiguration` as
```
Comment (Buzzardo) [inline README.adoc:92]: Add a colon to the end of the line.

File: README.adoc [hunk 18]
```diff
@@ -2,25 +2,151 @@
...
+
+====
+[source,java,tabsize=2,indent=0]
+----
+include::name-processor/src/main/java/com/example/stream/nameprocessor/NameProcessorConfiguration.java[tag=processname]
+----
+====
+
+This function will take a String value as input and create a new Person record that adds the timestamp of when the data was processed.
```
Comment (Buzzardo) [inline README.adoc:101]: Change "will take" to "takes" and "create" to "creates." (In other words, remove "will".)

File: README.adoc [hunk 19]
```diff
@@ -2,25 +2,151 @@
...
+Running this application will create 2 new exchanges in RabbitMQ: `processName-in-0` and `processName-out-0`.
```
Comment (Buzzardo) [inline README.adoc:102]: Change "will create 2" to "creates two".

File: README.adoc [hunk 20]
```diff
@@ -2,25 +2,151 @@
...
+Similar to the configuration we applied to the sink application, we want to change these exchange names so they can hook into the
+sink and soon to the supplier.
```
Comment (Buzzardo) [inline README.adoc:104]: Add commas around "soon".

File: README.adoc [hunk 21]
```diff
@@ -2,25 +2,151 @@
...
+====
+[source,tabsize=2,indent=0]
+----
+include::name-processor/src/main/resources/application.properties[tag=processorexchangeconfig]
+----
+====
+
+NOTE: The output of the processor matches the input of the sink.
+
+Using the RabbitMQ dashboard, we will now send a String (name) to the processor input exchange and watch as it flows to the connected sink.
```
Comment (Buzzardo) [inline README.adoc:115]: Change "will" to "can".

File: README.adoc [hunk 22]
```diff
@@ -2,25 +2,151 @@
...
+
+image::rabbitmq-processor.png[]
+
+If the processor and the sink are connected properly, you should still the output from your running sink:
```
Comment (Buzzardo) [inline README.adoc:119]: Change "still" to "see".

File: README.adoc [hunk 23]
```diff
@@ -2,25 +2,151 @@
...
+
+image::output-processor-sink.png[]
+
+== Source Application
+
+
+The Source (`java.util.function.Supplier`) is defined in `NameSourceConfiguration` as
```
Comment (Buzzardo) [inline README.adoc:126]: Add a colon to the end of the line.

File: README.adoc [hunk 24]
```diff
@@ -2,25 +2,151 @@
...
-them to Cloud Foundry, to Kubernetes, and on your local machine.
+====
+[source,tabsize=2,indent=0]
+----
+include::name-source/src/main/resources/application.properties[tag=supplierexchangeconfig]
+----
+====
 
-_(There is a single guide for both Apache Kafka and RabbitMQ - select the 'RabbitMQ' tabs)_.
+NOTE: The output of the source should matches the input of the processor
```
Comment (Buzzardo) [inline README.adoc:145]: Change "matches" to "match" and add a period at the end.

File: README.adoc [hunk 25]
```diff
@@ -2,25 +2,151 @@
...
 
-link:https://dataflow.spring.io/docs/stream-developer-guides/streams/standalone-stream-sample/[Stream Processing with RabbitMQ]
+If `name-processor` and `name-sink` are already running, starting the `name-source` will immediately start messages flowing through the system.
```
Comment (Buzzardo) [inline README.adoc:147]: Change "will immediately start" to "immediately starts". (In other words, remove "will".

File: README.adoc [hunk 26]
```diff
@@ -2,25 +2,151 @@
...
+You should see the same name continuously outputted by `name-sink` with a slightly different timestamp for when the message
```
Comment (Buzzardo) [inline README.adoc:148]: Change "outputted" to "produced".

(The past tense of "output" is "output", but I suggest avoiding the whole issue by using a different word.

**Review by Buzzardo — CHANGES_REQUESTED** [review]

**Review by Buzzardo — APPROVED** [review]

**Pull request merged**